extensions/resume.md

```diff
@@ -0,0 +1,258 @@
+    C2 - C: USER d * 0 :An example user!
+    C2 - S: :irc.example.com CAP * LS :multi-prefix draft/resume-0.3 sasl
+    C2 - C: CAP REQ :multi-prefix draft/resume-0.3 sasl
+    C2 - S: :irc.example.com CAP dan-backup-nick ACK :multi-prefix draft/resume-0.3 sasl
+    C2 - S: :irc.example.com RESUME TOKEN JKbAypzFiovffzuD8VEfcs6bOLrXsSenxsyZNt8
+    C2 - C: RESUME A8KgnZPYDaRiGMzZWLu2frVvtN7lbCxO3hTwGLO 2017-04-13T15:12:51.620Z
+    C2 - S: :irc.example.com RESUME SUCCESS dan
+    ... C1's connection is closed and C1's attributes are applied to C2 ...
+    C2 - S: :irc.example.com 001 dan :Welcome to the Internet Relay Network dan
+    ... C2 receives regular registration burst ...
+    C2 - S: :irc.example.com 376 dan :End of MOTD command
+    C2 - S: :dan!~d@127.0.0.1 JOIN #test
+    C2 - S: :irc.example.com 332 dan #test :Example topic
+    C2 - S: :irc.example.com 333 dan #test george 1442060874
+    C2 - S: :irc.example.com 353 dan @ #test :@dan @george +violet roger
+    C2 - S: :irc.example.com MODE #test +o dan
+
+Here is this successful reconnection seen by `george`, a client that does not have the `draft/resume-0.3` capability enabled:
+
+    S: :dan!~old@192.168.0.5 QUIT :Client reconnected (24 seconds of message history lost)
+    S: :dan!~d@127.0.0.1 JOIN #test
+    S: :irc.example.com MODE #test +o dan
+
+And here is this reconnection seen by `violet`, a client that has the `draft/resume-0.3` capability:
+
+    S: :dan!~old@192.168.0.5 RESUMED ~d 10.0.0.3 2017-04-13T15:12:51.620Z
+
+### Failed Resumption
+
+Failed attempt from a client with the nick `dan` reconnecting. The old connection used the username `~old` and the host `192.168.0.5`, and the new connection uses the username `~d` and the host `10.0.0.3`:
+
+    C1 - C: PING 12345678
+    C1 - S: :irc.example.com PONG 12345678
+    C1 - C: PING 87654321
+    ... C1 - Server does not send a response ...
+    ... C1 reconnects as C2 ...
+
+    C2 - C: CAP LS
+    C2 - C: NICK dan-backup-nick
+    C2 - C: USER d * 0 :An example user!
+    C2 - S: :irc.example.com CAP * LS :multi-prefix draft/resume-0.3 sasl
+    C2 - C: CAP REQ :multi-prefix draft/resume-0.3 sasl
+    C2 - S: :irc.example.com CAP dan-backup-nick ACK :multi-prefix draft/resume-0.3 sasl
+    C2 - S: :irc.example.com RESUME TOKEN JKbAypzFiovffzuD8VEfcs6bOLrXsSenxsyZNt8
+    C2 - C: RESUME A8KgnZPYDaRiGMzZWLu2frVvtN7lbCxO3hTwGLO 2017-04-13T15:12:51.620Z
+    C2 - S: :irc.example.com RESUME ERR :Cannot resume connection, token is not valid
+    C2 - C: AUTHENTICATE PLAIN
+    C2 - S: :irc.example.com AUTHENTICATE +
+    C2 - C: AUTHENTICATE YnVubnkAYnVubnkAYnVubnk=
+    C2 - S: :irc.example.com 900 dan-backup-nick * bunny :You are now logged in as bunny
+    C2 - S: :irc.example.com 903 dan-backup-nick :SASL authentication successful
+    C2 - S: :irc.example.com 001 dan-backup-nick :Welcome to the Internet Relay Network dan-backup-nick
+    ... regular connection ...
+
+
+## Implementation Considerations
```

Comment:
If client reconnects to a different server, the new server should be allowed to resume the connection, so the connection would be transferred to the new server. And they need to handle corner case when several clients try to resume the same connection simultaneously.

Reply:
I'd honestly prefer `resume` to require you connect to the same server again - just because i think server jumping is a bit too much like #330 and will be complex for ircd devs to implement

Reply:
I think it would be useful to have optional support for it in case for some reason you can't contact the old server.

Reply:
As a server maintainer I can't see any massive technical issues with implementing server jumping.

Reply:
Would it make sense to have a soft recommendation that clients attempt to reconnect to the same server initially (for compatibility with servers that don't synchronize resume tokens across nodes)?

Reply:
I thought that recommendation was already in the spec. If it's not, I'll add a note like that.

Reply:
Added guidance along both of these lines, for both server and client authors!